mountpoint-s3-crt/src/io/async_stream.rs

```diff
@@ -0,0 +1,246 @@
+//! AWS async input streams.
+
+use std::ptr::NonNull;
+
+use crate::common::{allocator::Allocator, error::Error};
+use async_channel::{RecvError, Sender};
+use log::debug;
+use mountpoint_s3_crt_sys::*;
+use thiserror::Error;
+
+/// Create an [AsyncInputStream] and its associated [AsyncStreamWriter].
+pub fn new_stream(allocator: &Allocator) -> (AsyncInputStream, AsyncStreamWriter) {
+    let (sender, receiver) = async_channel::bounded(1);
+    let stream = AsyncInputStream::new(allocator, sender);
+    let writer = AsyncStreamWriter { receiver };
+    (stream, writer)
+}
+
+/// An async input stream used to send the body of a request asynchronously.
+#[derive(Debug)]
+pub struct AsyncInputStream {
+    /// The inner aws_async_input_stream
+    pub(crate) inner: NonNull<aws_async_input_stream>,
+}
+
+impl AsyncInputStream {
+    fn new(allocator: &Allocator, sender: Sender<ReadRequest>) -> Self {
+        let ptr = Box::new(AsyncInputStreamImpl {
+            inner: Default::default(),
+            sender,
+        });
+
+        // Turn out Box pointer into a raw pointer (effectively leaking it).
+        let ptr = Box::into_raw(ptr) as *mut AsyncInputStreamImpl;
+
+        // SAFETY: We know ptr isn't null because we just made it from Box.
```

Comment:
`AsyncInputStreamImpl` becomes self-referential here (`ptr.inner.impl == ptr`). Usually that means we'd have to `Pin` the `Box<AsyncInputStreamImpl>`, but because that thing never escapes this module, we can just document it in this safety comment instead.

Actually, `async_input_stream_to_impl` should maybe return `Pin<&mut AsyncInputStreamImpl>` just to make this expectation clear.

To see why `Pin` is necessary, suppose we did this in `read_impl`, which compiles fine:
```rust
let b = Box::from_raw(async_input_stream_to_impl(stream));
let b = Box::new(*b);
let stream = Box::leak(b);
```

Now we've moved the `AsyncInputStreamImpl` to a different address, and freed the old one, so `ptr.inner.impl` is bogus and we'll probably crash the next time we try to use this. Pinning prevents us from making the move:
```rust
let b = std::pin::Pin::new_unchecked(Box::from_raw(async_input_stream_to_impl(stream)));
let b = Box::new(*b);  // <-- no longer compiles -- cannot move out of dereference of Pin 
let stream = Box::leak(b);
```